tasks/DroneFlightControlTask.cpp

```diff
@@ -0,0 +1,442 @@
+
+#include "DroneFlightControlTask.hpp"
+
+using namespace DJI::OSDK;
+using namespace drone_dji_sdk;
+using namespace VehicleStatus;
+
+DroneFlightControlTask::DroneFlightControlTask(std::string const &name)
+    : DroneFlightControlTaskBase(name)
+{
+}
+
+DroneFlightControlTask::~DroneFlightControlTask()
+{
+}
+
+/// The following lines are template definitions for the various state machine
+// hooks defined by Orocos::RTT. See DroneFlightControlTask.hpp for more detailed
+// documentation about them.
+
+bool DroneFlightControlTask::configureHook()
+{
+    if (!DroneFlightControlTaskBase::configureHook())
+        return false;
+
+    // Init class members
+    mFunctionTimeout = 1; // second
+    mStatusFreqInHz = 10; // Hz
+    // waypoint init settings
+    mMission.max_velocity = 10;
+    mMission.idle_velocity = 5;
+    mMission.latitude.rad = 0;
+    mMission.longitude.rad = 0;
+    mMission.altitude = 0;
+    mMission.finish_action = FinishAction::NO_FINISH_ACTION;
+    mMission.executive_times = ExecTimes::ONCE;
+    mMission.yaw_mode = YawMode::YAW_AUTO_MODE;
+    mMission.trace_mode = TraceMode::POINT_TO_POINT;
+    mMission.rc_lost_action = RcLostAction::CONTINUE_WAYPOINT;
+    mMission.gimbal_pitch = GimbalPitch::FREE_MODE;
+
+    // Configure GPS stuffs
+    mGPSSolution.setParameters(_gps_property.get());
+
+    if (!initVehicle())
+        return false;
+    if (!checkTelemetrySubscription())
+        return false;
+    if (!missionInitSettings())
+        return false;
+
+    return true;
+}
+
+bool DroneFlightControlTask::startHook()
+{
+    if (!DroneFlightControlTaskBase::startHook())
+        return false;
+
+    // Obtain Control Authority
+    mVehicle->obtainCtrlAuthority(mFunctionTimeout);
+    return true;
+}
+
+typedef DroneFlightControlTask::States TaskState;
+static TaskState djiStatusFlightToTaskState(uint8_t status)
+{
+    switch (status)
+    {
+    case DJI::OSDK::VehicleStatus::FlightStatus::STOPED:
+        return TaskState::DJI_STOPPED;
+    case DJI::OSDK::VehicleStatus::FlightStatus::ON_GROUND:
+        return TaskState::DJI_ON_GROUND;
+    case DJI::OSDK::VehicleStatus::FlightStatus::IN_AIR:
+        return TaskState::DJI_IN_AIR;
+    }
+    // Never reached
+    throw std::invalid_argument("invalid controller state");
+}
+
+void DroneFlightControlTask::updateHook()
+{
+    _pose_samples.write(getRigidBodyState());
+    _battery.write(getBatteryStatus());
+
+    // cmd input
+    CommandAction cmd_input;
+    if (_cmd_input.read(cmd_input) == RTT::NoData)
+        return;
+
+    // setpoint input
+    VehicleSetpoint cmd_pos;
+    if (_cmd_pos.read(cmd_pos) != RTT::NewData)
+        return;
+
+    // Check status
+    auto djiStatusFlight = mVehicle->subscribe->getValue<Telemetry::TOPIC_STATUS_FLIGHT>();
+    DroneFlightControlTask::States status = djiStatusFlightToTaskState(djiStatusFlight);
+    if (state() != status)
+        state(status);
```

Comment:
I think we should put this before the cmd_input read, with the rest of the state feedback